Fix | Build DataTypes table in code, add `json` data type

edwardneal opened the pull request:
## Description

This PR makes two related changes: it populates the `DataTypes` table returned by `SqlConnection.GetSchema` in code rather than by deserializing an embedded XML resource, and then adds the `json` data type to this table.

In the process of adding tests, I also noticed that there wasn't a coherent view of whether or not the `json` data type was present, so added one. This also led me to notice that we're not running JSON-based tests on SQL 2025, so I changed their test criteria to make sure this happens. This is essentially the same situation as #3794, and I'm expecting some degree of conflicts depending upon the order they're merged in.

It should be possible to move commit-by-commit, but 2f907edb6481597d364a5c0220de083be10fa124 is probably the bulkiest. This commit populates the fields in the `DataTypes` table based on the value of fields in the `MetaType` instances for each type.

## Issues

Contributes to #3833.
Contributes to #3372.

I'm conscious that #3833 refers to the `json` and the `vector` types. I've not included the vector type because I don't know whether we should treat `vector({0}[, float32])` and `vector({0}, float16)` as two different data types. There's also a little extra plumbing to deal with within `SqlVector<float>` and in the tests (which #3794 introduces.)

## Testing

I've manually verified that all fields for all existing data types are identical between main and this branch. I've also added a new test to prove that no data types have been omitted.

edwardneal committed:
Create structure of DataTypes static population

edwardneal committed:
Remove redundant data types from metadata

SqlClient cannot connect to SQL Server 2000 instances, so the minimum server version is 9.0.
* Remove data type entries with a maximum version below this.
* Remove the minimum version constraint where this constraint will always be true.

edwardneal committed:
Populate DataTypes table from MetaType information

edwardneal committed:
Add test to verify presence of data types

edwardneal committed:
Add json data type to DataTypes schema collection

edwardneal committed:
Use IsJsonSupported more widely

This enables the JSON tests to run against SQL Server 2025.
Also correct a comment in LoadDataTypesDataTables.

edwardneal committed:
Plumb minimum version logic to AddStringOrBinaryType

paulmedynski commented on src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/DataBaseSchemaTest/ConnectionSchemaTest.cs:
```diff
@@ -107,29 +108,91 @@ public static void GetStructuredTypeMembersFromSchema()
+            Assert.All(expectedTypes, type => Assert.Contains(type, actualTypes));
+
+            // The "json" type should only be present when running against a SQL Server version which supports it.
+            // SQL Azure reports a version of 12.x but supports JSON, so SqlClient doesn't include it in the list of types.
```
Azure SQL versions aren't comparable with SQL Server versions, AFAICT.  Azure SQL uses Engine Edition 5, so we can check for that and assume all Azure SQL instances support JSON.  There are specific version numbers that we may be able to go back and confirm when JSON support first appeared for Azure SQL, but I don't think it's worth it.

Same idea for vectors, except that underlying data types will be added as time goes on, and we will only know by performing the login handshake.

edwardneal replied:
I agree, but it's a wider design issue with SqlMetaDataFactory. It decides whether a row should appear in the final DataTable based upon the reported server version (in [`SupportedByCurrentVersion`](https://github.com/dotnet/SqlClient/blob/main/src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlMetaDataFactory.cs#L51).) In theory, if we decided to use a different query for one of the other metadata collections (e.g. Tables) or to add a new metadata collection which only appeared in SQL Server 2016 / v13.x or higher, the filtering logic would hide it for Azure SQL instances.

We need a way to convert feature IDs and versions to a list of server capabilities (i.e. JSON feature with v1, vector feature with float32, vector feature with float16, etc.) and to allow filtering on those. That's capability list is best placed within SqlInternalConnectionTds though and this PR's already pretty large, so I'd prefer to do that separately prior to preview4.

paulmedynski replied:
Sound good - I just wanted to point the different way the version numbers are treated for the different engine editions.

paulmedynski commented on src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlMetaDataFactory.DataTypes.cs:
Are there existing unit tests for this functionality that are now testing this new file?  Or do we need new tests for this?

edwardneal committed:
Merge main

paulmedynski commented:
@edwardneal - New conflicts here.

edwardneal committed:
Merge main

apoorvdeshmukh commented:
/azp run

edwardneal commented:
Thanks @paulmedynski and @apoorvdeshmukh. I think my comment may be a bit redundant, but... I've just merged!

apoorvdeshmukh reviewed (changes requested):
@edwardneal There are some test failures. Check the testcase `TestJsonWithMARS`

edwardneal committed:
Address MARS JSON test failure

This was manually constructing a connection string, which didn't work on one CI leg.

apoorvdeshmukh reviewed (approved):
LGTM
